cpp/unit/selection.cpp

```diff
@@ -103,17 +104,92 @@ void UnitSelection::clear() {
 		}
 	}
 	this->units.clear();
+	this->selection_type = selection_type_t::nothing;
 }
 
-void UnitSelection::toggle_unit(Unit *u) {
-	if (this->units.count(u->id) > 0) {
-		u->selected = false;
-		this->units.erase(u->id);
+void UnitSelection::toggle_unit(Unit *u, bool append) {
+	if (this->units.count(u->id) == 0) {
+		this->add_unit(u, append);
+	} else {
+		this->remove_unit(u);
 	}
-	else {
-		u->selected = true;
-		this->units[u->id] = u->get_ref();
+}
+
+void UnitSelection::add_unit(Unit *u, bool append) {
+	// Dead units aren't selectable
+	if (u->has_attribute(attr_type::hitpoints) && u->get_attribute<attr_type::hitpoints>().current <= 0) {
+		return;
+	}
+
+	selection_type_t unit_type = get_unit_selection_type(u);
+	int unit_type_i = static_cast<int>(unit_type);
+	int selection_type_i = static_cast<int>(this->selection_type);
+
+	if (unit_type_i > selection_type_i) {
+		// Don't select this unit as it has too low priority
+		return;
+	}
+
+	if (unit_type_i < selection_type_i) {
+		// Upgrade selection to a higher priority selection
+		this->units.clear();
+		this->selection_type = unit_type;
 	}
+
+	// Can't select multiple enemies at once
+	if (not (unit_type == selection_type_t::own_units ||
+	        (unit_type == selection_type_t::own_buildings && append))) {
+
+		this->clear();
+		this->selection_type = unit_type; // Clear resets selection_type
+	}
+
+	// Finally, add the unit to the selection
+	u->selected = true;
+	this->units[u->id] = u->get_ref();
+}
+
+void UnitSelection::remove_unit(Unit *u) {
+	u->selected = false;
+	this->units.erase(u->id);
+
+	if (this->units.empty()) {
+		this->selection_type = selection_type_t::nothing;
+	}
+}
+
+void UnitSelection::kill_unit() {
+	if (this->units.empty()) {
+		return;
+	}
+
+	UnitReference &ref = this->units.begin()->second;
+	if (!ref.is_valid()) {
+		this->units.erase(this->units.begin());
+
+		if (this->units.empty()) {
+			this->selection_type = selection_type_t::nothing;
+		}
+	} else {
+		Unit *u = ref.get();
+
+		// Check color: you can only kill your own units
+		int player = Engine::get().current_player;
+		int color = u->get_attribute<attr_type::owner>().player.color;
+		if (player == color) {
+			this->remove_unit(u);
+			u->delete_unit();
+		}
+	}
+}
+
+bool UnitSelection::contains_villagers() {
+	for (auto &it : units) {
+		if (it.second.is_valid() && it.second.get()->get_ability(ability_type::build)) {
+			return true;
+		}
```

Comment:
sure that you want to depend "contains villavers" on the build ability? maybe should be `contains_builders`?